utils/error_injection.hh

```diff
@@ -534,11 +534,8 @@ public:
 
     // \brief Inject exception
     // \param func function returning a future and taking an injection handler
-    template <typename Func>
-    requires std::is_invocable_r_v<future<>, Func, error_injection<true>::injection_handler&>
     [[gnu::always_inline]]
-    future<> inject(const std::string_view& name,
-                                 Func&& func) {
+    future<> inject(const std::string_view& name, waiting_handler_fun func) {
```

Comment:
This causes creation of an std::function, then immediate destruction. std::function can allocate if the lambda captures many parameters (typically I expect it will capture none). Please verify it is optimized to nothing.

If not, it should remain a template.

Reply:
btw for the empty case, the inject template is easier to instantiate than the internal stuff std::function instantiates.

Reply:
Compiled
```c++
future<> inject_empty_capture() {
    return utils::get_local_injector().inject("foo", [] (auto& handler) {
        return handler.wait_for_message(db::timeout_clock::now() + std::chrono::seconds(1));
    });
}

struct large {
    int x[32];
};

future<> inject_large_capture(large arg) {
    return utils::get_local_injector().inject("bar", [arg] (auto& handler) {
        if (arg.x[arg.x[0] % 32] == arg.x[arg.x[1] % 32]) {
            return handler.wait_for_message(db::timeout_clock::now() + std::chrono::seconds(1));
        } else {
            return make_ready_future<>();
        }
    });
}
```

This PR (std::function):
```
000000000474e710 <_Z20inject_empty_capturev>:
 474e710:	48 89 f8             	mov    %rdi,%rax
 474e713:	48 c7 07 00 00 00 00 	movq   $0x0,(%rdi)
 474e71a:	48 c7 47 08 03 00 00 	movq   $0x3,0x8(%rdi)
 474e721:	00 
 474e722:	c3                   	ret
 474e723:	cc                   	int3
 ...

000000000474e730 <_Z20inject_large_capture5large>:
 474e730:	48 89 f8             	mov    %rdi,%rax
 474e733:	48 c7 07 00 00 00 00 	movq   $0x0,(%rdi)
 474e73a:	48 c7 47 08 03 00 00 	movq   $0x3,0x8(%rdi)
 474e741:	00 
 474e742:	c3                   	ret
 474e743:	cc                   	int3
 ...
```

Master (template):
```
0000000004742560 <_Z20inject_empty_capturev>:
 4742560:	48 89 f8             	mov    %rdi,%rax
 4742563:	48 c7 07 00 00 00 00 	movq   $0x0,(%rdi)
 474256a:	48 c7 47 08 03 00 00 	movq   $0x3,0x8(%rdi)
 4742571:	00 
 4742572:	c3                   	ret
 4742573:	cc                   	int3
 ...
0000000004742580 <_Z20inject_large_capture5large>:
 4742580:	48 89 f8             	mov    %rdi,%rax
 4742583:	48 c7 07 00 00 00 00 	movq   $0x0,(%rdi)
 474258a:	48 c7 47 08 03 00 00 	movq   $0x3,0x8(%rdi)
 4742591:	00 
 4742592:	c3                   	ret
 4742593:	cc                   	int3
 ...
```

Reply:
> btw for the empty case, the inject template is easier to instantiate than the internal stuff std::function instantiates.

It can as well be noncopyable_function if we're more confident in it :)

Reply:
> Please verify it is optimized to nothing.

@avikivity , is the above verification reasonable?

Reply:
yes